core/src/main/scala/scalaz/IList.scala

```diff
@@ -467,24 +483,61 @@ sealed abstract class IList[A] extends Product with Serializable {
   def zipWithIndex: IList[(A, Int)] =
     zip(IList(0 until length : _*))
 
+
+  final override def productElement(n: Int): Any = value.list.productElement(n)
+
+  final override def productArity: Int = value.list.productArity
+
+  final override def canEqual(that: Any): Boolean = that.isInstanceOf[IList[A]]
+
+  final override def equals(that: Any): Boolean = canEqual(that) && that.asInstanceOf[IList[A]].value.equals(value)
+
 }
 
-// In order to get exhaustiveness checking and a sane unapply in both 2.9 and 2.10 it seems
-// that we need to use bare case classes. Sorry. Suggestions welcome.
-final case class INil[A]() extends IList[A]
-final case class ICons[A](head: A, tail: IList[A]) extends IList[A]
+sealed abstract class StrictIList[A] {
+  def list: IList[A]
+}
+sealed abstract case class INil[A]() extends StrictIList[A] { self: IList[A] =>
+  override def value: StrictIList[A] = this
+  override def list: IList[A] = this
+}
+object INil {
+  private[this] val nil: IList[Nothing] = new INil with IList[Nothing] {}
+  def apply[A](): IList[A] = nil.asInstanceOf[IList[A]]
+}
+sealed abstract case class ICons[A](head: A, tail: IList[A]) extends StrictIList[A] { self: IList[A] =>
+  override def value: StrictIList[A] = this
+  override def list: IList[A] = this
+}
+object ICons {
+  def apply[A](head: A, tail: IList[A]): IList[A]  = new ICons[A](head, tail) with IList[A] {}
+}
+private final class ByNeed[A](private[this] var list: () => IList[A]) extends IList[A] {
+  override lazy val value = {
+    val value0 = list().value
```

Comment:
If I implement `filterLazy` using `flatMapLazy` and then filter out too many elements in a row.